src/null-safety/migration-guide.md

```diff
@@ -1,23 +1,416 @@
+4. If the package is already on pub.dev,
+   **publish** the null-safe version as a **prerelease** version.
 
-[null safety homepage]: /null-safety
-[Dart announcements]: {{site.group}}/d/forum/announce
-[Dart blog]: https://medium.com/dartlang
+{{ site.alert.info }}
+  **Migrating an app is the same as migrating a package.**
+  Before fully migrating an app,
+  consider waiting until null safety is in a stable release.
+  In the meantime, you can add migration hints to your app
+  and take the opportunity to refactor your app code.
+  **_[PENDING: is that good/helpful advice?]_**
+{{ site.alert.end }}
 
+[null safety]: /null-safety
+
+![Screenshot of migration tool](/null-safety/migration-tool.png)
+_This interactive tool (`dart migrate`) can simplify migration to null safety._
+
+**_[PENDING: flutter-specific stuff?
+Only things kevmoo's run into:
+--[no-]sound-null-safety (probably applies to `dart` as well)
+IDE launch options
+flutter run, flutter test flag order might be slightly different.
+pub run --enable-experiment=non-nullable test
+aka dart test --enable-experiment=non-nullable]_**
+
+## 1. Wait to migrate {#step1-wait}
+
+We strongly recommend migrating code in order, 
+with the leaves of the dependency graph being migrated first.
+For example, if package C depends on package B, which depends on package A,
+then A should be migrated to null safety first, then B, then C.
+
+![Illustration of C/B/A sentence](/null-safety/null-safety-migration-order.png){:width="454px"}<br>
+
+Although you _can_ migrate before your dependencies support null safety,
+you might have to change your code when your dependencies migrate.
+For example, if you predict that a function will take a nullable parameter but
+the package migrates it to be non-nullable,
+then passing a nullable argument becomes a compile error.
+
+
+### Switch to a beta release
+
+Start the migration process by switching to the **latest beta release**
+of either the Dart SDK or the Flutter SDK,
+and then updating your pubspec to match.
+
+How you get the latest beta release depends on whether
+you use the Flutter SDK:
+
+* If you use the Flutter SDK,
+  switch to the beta channel:
+
+  ```terminal
+$ flutter channel beta
+$ flutter upgrade
+```
+* Otherwise, download a beta release
+  from the [Dart SDK archive][].
+
+[Dart SDK archive]: /tools/sdk/archive#beta-channel
+
+After updating your release,
+update your pubspec to the version of Dart you're using.
+For example, say you're using 2.12.0-18.0.beta:
+**[PENDING: check. Is this really necessary? It caused me grief when I tried to run `dart migrate`.]**
+
+```terminal
+$ dart --version
+Dart SDK version: 2.12.0-18.0.beta (beta)... [PENDING: check!]
+```
+
+Update the `pubspec.yaml` file to have that version as a lower constraint:
+**[PENDING: check!!!]**
+
+```yaml
+environment:
+  sdk: '>=2.12.0-18.0.beta <3.0.0'
+```
+
+
+{{ site.alert.warn }}
+  When you return to working on production code, remember to
```

Comment:
But that will downgrade their Dart SDK and make any null safety edits they've made turn into compiler errors..

Reply:
I changed it to "If". The intent was just to remind them, especially if they're just checking on the status of dependencies and not in the midst of really migrating.